### PR TITLE
HyperConformer

### PR DESCRIPTION
**Pull request by florianmai**
I added an implementation for the HyperConformer architecture.
In short, HyperConformer replaces the self-attention module in Conformer with multi-head HyperMixer, an extension of [the original HyperMixer](https://arxiv.org/abs/2203.03691) proposed for NLP. Since HyperMixer has linear memory and time complexity in the sequence length, it is an efficient alternative to Conformer while attaining similar or sometimes better results.

Regarding the implementation, I made three major changes. First, I added the "hypermixing.py" module to nnet. Second, I made minimal changes to lobes/models/ASR/transformer/Transformer /Conformer and /TransformerASR such that HyperMixing is supported as a type of attention. Third, I added two new recipes for training HyperConformer on LibriSpeech.

### CODE REVIEW

**Comment by mravanelli** [comment]
Thank you Florian,
do you already have some performance numbers?

**Comment by JuanPZuluaga** [comment]
Hi @mravanelli and @Adel-Moumen we have on test-clean + LM (same for conformer/transformer): ~2.31% WER w/ ~22M params, and ~2.86% WER w/ 7.8M params.

**Comment by florianmai** [comment]
> Hi @mravanelli and @Adel-Moumen
> 
> We got on test-clean + LM (same LM used for conformer/transformer): ~2.31% WER w/ ~22M params, and ~2.86% WER w/ 7.8M params.

In a controlled setting, a 24M parameter Conformer reaches 2.6% WER and an 8.7M parameter Conformer reaches 3.1%. HyperConformer uses 20% and 30% less memory at the peak, respectively.

We are also working on a ~100M parameter model, but tuning it takes some time.

**Comment by JuanPZuluaga** [comment]
Hello @Adel-Moumen, we just added the link to the models trained with HyperConformer. Could you please check this PR? 

Thanks.

**Comment by Adel-Moumen** [comment]
Hello @JuanPZuluaga and @florianmai, 

Thanks again for your great PR. It is a really nice addition to SpeechBrain. 

This PR will be reviewed by @mhn226 and @TParcollet. But I can help if you do have any questions (here or on twitter/discord)! :)

File: recipes/LibriSpeech/ASR/transformer/README.md [hunk 1]
```diff
@@ -20,7 +20,9 @@ installed in your environment (see extra-requirements.txt)**
 |:-------------:|:---------------------------:| :-----:| :-----:| :-----:| :--------:|
 | 24-03-22 | transformer.yaml | 2.27 | [HuggingFace](https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech) | [GoogleDrive](https://drive.google.com/drive/folders/1Nv1OLbHLqVeShyZ8LY9gjhYGE1DBFzFf?usp=sharing) | 4xV100 32GB |
 | 24-03-22 | conformer_small.yaml | 2.49 (**only 13.3M parameters**) | [HuggingFace](https://huggingface.co/speechbrain/asr-conformersmall-transformerlm-librispeech) | [GoogleDrive](https://drive.google.com/drive/folders/1I4qntoodHCcj1JNbDrfwFHYcLyu1S5-l?usp=sharing) | 1xV100 32GB |
-| 06-12-23 | train_hf_whisper.yaml | 3.60 | Not Avail. | Not Avail. | 1xA100 40GB |
+| 06-12-22 | train_hf_whisper.yaml | 3.60 | Not Avail. | Not Avail. | 1xA100 40GB |
+| 27-03-23 | train_hyperconformer_8M.yaml | 2.55  (**only 7.9M parameters**) | NA | [GoogleDrive](https://drive.google.com/drive/folders/1XR0DLQJHLG5U3MWP1HDvqUvc4VqM60dq?usp=drive_link) | 1xP40 24GB |
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/README.md:24]: I will upload your google drive link to DropBox. I will give you the new link asap to change this one. Thanks again for releasing the weights.

File: recipes/LibriSpeech/ASR/transformer/README.md [hunk 2]
```diff
@@ -20,7 +20,9 @@ installed in your environment (see extra-requirements.txt)**
...
+| 27-03-23 | train_hyperconformer_22M.yaml | 2.23 (**only 21.7M parameters**) | NA | [GoogleDrive](https://drive.google.com/drive/folders/1ixH8vHeWR_QsPlhZfjWSxctSyuxgv8y2?usp=drive_link) | 1xP40 24GB |
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/README.md:25]: I will upload your google drive link to DropBox. I will give you the new link asap to change this one. Thanks again for releasing the weights.

File: recipes/LibriSpeech/ASR/transformer/README.md
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/README.md:1]: you should add a section regarding hyper conformer, explaining a little bit what it is and how to cite your work! :)

**Review by Adel-Moumen — CHANGES_REQUESTED** [review]
Hello guys!

Thanks for your great work. This is a nice addition to SpeechBrain. 

In the end, I'll be the one doing the review for different reasons. I leave a couple of comments regarding your work. My first comment is regarding TransformerLM. Why are you not using it? We should just remove all the TransformerLM references if the LM is not part of the recipe. 

I will shortly upload your Google Drive models on DropBox so that it is `wget` compliant. Is it possible to add your models on the HuggingFace Hub? This is a great way to advertise a bit more your work (see: https://huggingface.co/speechbrain). If you do agree, please join the SpeechBrain org, I will accept you and grant you with the contributor access. Have a look at our Conformer/Transformer model files to understand a little bit how it works (not that hard believe me!). 

Please add your recipe in our test recipe csv: 
https://github.com/speechbrain/speechbrain/blob/develop/tests/recipes/LibriSpeech.csv 

One last thing, please fetch the latests commits from the SpeechBrain dev branch! :)

This way we can run our testing suits on your recipe. 

Overall as I said, this is a great work. I just have one concern regarding where you put your HyperMixer. I don't think this is the right place. (CC: @TParcollet @mravanelli)

File: recipes/LibriSpeech/ASR/transformer/README.md (hunk 1 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/README.md:24]: https://www.dropbox.com/sh/8jc96avmivr8fke/AABrFEhtWy_3-Q7BHhkh0enwa?dl=0

File: recipes/LibriSpeech/ASR/transformer/README.md (hunk 2 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/README.md:25]: https://www.dropbox.com/sh/30xsmqj13jexzoh/AACvZNtX1Fsr0Wa1Z3C9rHLXa?dl=0

File: recipes/LibriSpeech/ASR/transformer/README.md (hunk 1 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/README.md:24]: Please make sure that everything is in the DropBox (to double check)

File: recipes/LibriSpeech/ASR/transformer/README.md (hunk 2 above)
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/README.md:25]: Please make sure that everything is in the DropBox (to double check)

File: recipes/LibriSpeech/ASR/transformer/hparams/hyperconformer_22M.yaml [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+train_log: !ref <output_folder>/train_log.txt
+
+# Language model (LM) pretraining
+# NB: To avoid mismatch, the speech recognizer must be trained with the same
+# tokenizer used for LM training. Here, we download everything from the
+# speechbrain HuggingFace repository. However, a local path pointing to a
+# directory containing the lm.ckpt and tokenizer.ckpt may also be specified
+# instead. E.g if you want to use your own LM / tokenizer.
+# pretrained_lm_tokenizer_path: speechbrain/asr-transformer-transformerlm-librispeech
+pretrained_lm_tokenizer_path: !PLACEHOLDER  # e.g., /path/to/tokenizer/5K_subword_unigram_960h_LM
```
[the remark this answers is not in the document]
Comment (JuanPZuluaga) [reply recipes/LibriSpeech/ASR/transformer/hparams/hyperconformer_22M.yaml]: Actually, we are using TransformerLM. Sorry that we did not specify that before. Now it's fixed

[the remark this answers is not in the document]
Comment (Adel-Moumen) [reply recipes/LibriSpeech/ASR/transformer/hparams/hyperconformer_22M.yaml]: Ok, good. :)

**Comment by JuanPZuluaga** [comment]
Hi folks, I think I addressed most of the comments. I also added support for Branchformer + HyperMixer = HyperBranchformer. 

I haven't tested it yet. But it should work out of the box.

**Review by Adel-Moumen — CHANGES_REQUESTED** [review]
Hey,

Good job! It starts to look very good. 

However, there's still some small things to change. 

Coud you please add a new line for each new hparams in this file: https://github.com/speechbrain/speechbrain/blob/91cac2feee51febcbc6a34f7675234baac6ad1a6/tests/recipes/LibriSpeech.csv

Furthermore, please run the recipe tests:
```bash
cd speechbrain 
find recipes | grep extra | xargs cat | sort -u | grep -v \# | xargs -I {} pip install {}
tests/.run-recipe-tests.sh
```

Thanks again!

**Comment by JuanPZuluaga** [comment]
Hi @Adel-Moumen, all the fixes are incorporated now

File: recipes/LibriSpeech/ASR/transformer/README.md [hunk 4]
```diff
@@ -29,6 +29,30 @@ installed in your environment (see extra-requirements.txt)**
 | 24-03-22 | transformer.yaml | 3.32 | 2.27 | 5.53 | [HuggingFace](https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech) | [GoogleDrive](https://www.dropbox.com/sh/653kq8h2k87md4p/AAByAaAryXtQKpRzYtzV9ih5a?dl=0) | 4xV100 32GB |
 | 24-03-22 | conformer_small.yaml | 4.05 | 2.49 | 6.1 (**only 13.3M parameters**) | [HuggingFace](https://huggingface.co/speechbrain/asr-conformersmall-transformerlm-librispeech) | [GoogleDrive](https://www.dropbox.com/sh/s0x6ni124858b8i/AAALaCH6sGTMRUVTjh8Tm8Jwa?dl=0) | 1xV100 32GB |
 | 06-12-23 | train_hf_whisper.yaml | 3.60 | Not Avail. | Not Avail. | Not Avail. | Not Avail. | 1xA100 40GB |
+| 27-03-23 | hyperconformer_8M.yaml | 4.69 | 2.55 | 6.61 (**only 7.9M parameters**) | NA |  [DropBox](https://www.dropbox.com/sh/8jc96avmivr8fke/AABrFEhtWy_3-Q7BHhkh0enwa?dl=0) | 1xP40 24GB
+| 27-03-23 | hyperconformer_22M.yaml | 3.19 | 2.23 | 5.54  (**only 21.7M parameters**)  | NA | [DropBox](https://www.dropbox.com/sh/30xsmqj13jexzoh/AACvZNtX1Fsr0Wa1Z3C9rHLXa?dl=0) | 1xP40 24GB
+| 03-09-23 | hyperbranchformer_13M.yaml | NA | 2.54 | 6.58  | NA | soon | 1xP40 24GB
```
Comment (Adel-Moumen) [inline recipes/LibriSpeech/ASR/transformer/README.md:34]: We should take this opportunity to add the hyperbranchformer ckpt as well :)

File: tests/recipes/LibriSpeech.csv [hunk 5]
```diff
@@ -32,3 +32,5 @@ ASR,LibriSpeech,recipes/LibriSpeech/ASR/CTC/train_with_whisper.py,recipes/LibriS
...
+ASR,LibriSpeech,recipes/LibriSpeech/ASR/transformer/train.py,recipes/LibriSpeech/ASR/transformer/hparams/hyperconformer_8M.yaml,recipes/LibriSpeech/ASR/transformer/librispeech_prepare.py,recipes/LibriSpeech/ASR/transformer/README.md,,,--data_folder=tests/samples/ASR/ --train_csv=tests/samples/annotation/ASR_train.csv --valid_csv=tests/samples/annotation/ASR_train.csv --test_csv=[tests/samples/annotation/ASR_train.csv] --number_of_epochs=10 --skip_prep=True,"file_exists=[env.log,hyperparams.yaml,log.txt,train_log.txt,train.py,wer_ASR_train.txt,save/lm.ckpt,save/tokenizer.ckpt] performance_check=[train_log.txt, train loss, <500, epoch: 10]"
```
Comment (Adel-Moumen) [inline tests/recipes/LibriSpeech.csv:36]: as you added branchformer we also need it here :)

**Review by Adel-Moumen — CHANGES_REQUESTED** [review]
It start to looks good! some small things regarding hyperbranchformer to change but overall it is good. 

I will now play with the PR. 

Are you going to release the pretrained models on the HF hub?

File: recipes/LibriSpeech/ASR/transformer/README.md (hunk 4 above)
Comment (TParcollet) [reply recipes/LibriSpeech/ASR/transformer/README.md:34]: @Adel-Moumen not mandatory if this is blocking the merge, let's merge and update.

**Review by Adel-Moumen — APPROVED** [review]
All tests passed.

```
(1/33) Running test for LibriSpeech_row_02...
        ... 154.40s
(2/33) Running test for LibriSpeech_row_03...
        ... 34.54s
(3/33) Running test for LibriSpeech_row_04...
        ... 189.59s
(4/33) Running test for LibriSpeech_row_05...
        ... 157.21s
(5/33) Running test for LibriSpeech_row_06...
        ... 62.48s
(6/33) Running test for LibriSpeech_row_07...
        ... 61.20s
(7/33) Running test for LibriSpeech_row_08...
        ... 21.61s
(8/33) Running test for LibriSpeech_row_09...
        ... 51.68s
(9/33) Running test for LibriSpeech_row_10...
        ... 131.59s
(10/33) Running test for LibriSpeech_row_11...
        ... 191.26s
(11/33) Running test for LibriSpeech_row_12...
        ... 175.70s
(12/33) Running test for LibriSpeech_row_13...
        ... 157.18s
(13/33) Running test for LibriSpeech_row_14...
        ... 49.47s
(14/33) Running test for LibriSpeech_row_15...
        ... 57.72s
(15/33) Running test for LibriSpeech_row_18...
        ... 446.05s
(16/33) Running test for LibriSpeech_row_19...
        ... 171.78s
(17/33) Running test for LibriSpeech_row_20...
        ... 45.35s
(18/33) Running test for LibriSpeech_row_21...
        ... 8.87s
(19/33) Running test for LibriSpeech_row_22...
        ... 22.87s
(20/33) Running test for LibriSpeech_row_23...
        ... 15.90s
(21/33) Running test for LibriSpeech_row_28...
        ... 3.64s
(22/33) Running test for LibriSpeech_row_29...
        ... 3.65s
(23/33) Running test for LibriSpeech_row_30...
        ... 12.40s
(24/33) Running test for LibriSpeech_row_31...
        ... 17.59s
(25/33) Running test for LibriSpeech_row_32...
        ... 18.08s
(26/33) Running test for LibriSpeech_row_33...
        ... 198.02s
(27/33) Running test for LibriSpeech_row_34...
        ... 204.22s
(28/33) Running test for LibriSpeech_row_35...
        ... 41.32s
(29/33) Running test for LibriSpeech_row_36...
        ... 25.33s
(30/33) Running test for LibriSpeech_row_37...
        ... 42.05s
(31/33) Running test for LibriSpeech_row_38...
        ... 21.83s
(32/33) Running test for full_inference_row_02...
        download_file('https://www.openslr.org/resources/12/test-clean.tar.gz', 'tests/download/test-clean.tar.gz', unpack=True, dest_unpack='tests/download/LibriSpeech',write_permissions=True)
tests/download/test-clean.tar.gz exists. Skipping download
Extracting tests/download/test-clean.tar.gz to tests/download/LibriSpeech
        download_file('https://www.dropbox.com/sh/wy6ktfdxqt0dmou/AABY4ty3zy3xPef0zwRx949Oa?dl=1', 'tests/download/LibriSpeech_ASR_transformer.zip', unpack=True, dest_unpack='tests/download/LibriSpeech_ASR_transformer', write_permissions=True)
tests/download/LibriSpeech_ASR_transformer.zip exists. Skipping download
Extracting tests/download/LibriSpeech_ASR_transformer.zip to tests/download/LibriSpeech_ASR_transformer
                Running an Inference Test.
        ... 37.67s
(33/33) Running test for full_inference_row_03...
        download_file('https://www.openslr.org/resources/12/test-clean.tar.gz', 'tests/download/test-clean.tar.gz', unpack=True, dest_unpack='tests/download/LibriSpeech', write_permissions=True)
tests/download/test-clean.tar.gz exists. Skipping download
Extracting tests/download/test-clean.tar.gz to tests/download/LibriSpeech
        download_file('https://www.dropbox.com/sh/vs1os5mwh4necp0/AAAs7rMl4IO4N4s__BbVon6qa?dl=1', 'tests/download/LibriSpeech_ASR_seq2seq.zip', unpack=True, dest_unpack='tests/download/LibriSpeech_ASR_seq2seq', write_permissions=True)
tests/download/LibriSpeech_ASR_seq2seq.zip exists. Skipping download
Extracting tests/download/LibriSpeech_ASR_seq2seq.zip to tests/download/LibriSpeech_ASR_seq2seq
                Running an Inference Test.
        ... 29.78s
TEST PASSED
```

LGTM.

**Comment by TParcollet** [comment]
LGTM! Thank you everyone! especially @florianmai and @JuanPZuluaga

**Comment by JuanPZuluaga** [comment]
Thank you!! Really happy to have our first contribution. We will update the links with the `HyperBranchformer` models soon.

**Comment by Adel-Moumen** [comment]
Well done @JuanPZuluaga and @florianmai. Very good work.

**Pull request merged**